APE22.rst

```diff
@@ -0,0 +1,440 @@
+:ref:`description_of_the_problem`.  It is also the most authoritative source, as
+it is the official Python style guide.  However, it is not as unambiguous as we
+require: which takes precedence, the ``__all__`` attribute or the underscore
+prefix?
+
+`SciPy
+<https://docs.scipy.org/doc/scipy/reference/index.html#importing-from-scipy>`_
+tries to resolve this ambiguity by stating that ::
+
+   - Methods / functions / classes and module attributes whose names begin with
+     a leading underscore are private.
+   - If a class name begins with a leading underscore, none of its members are
+     public, whether or not they begin with a leading underscore.
+   - If a module name in a package begins with a leading underscore none of its
+     members are public, whether or not they begin with a leading underscore.
+   - If a module or package defines ``__all__``, that authoritatively defines
+     the public interface.
+   - If a module or package doesn’t define ``__all__``, then all names that
+     don’t start with a leading underscore are public.
+
+This is a good solution, and is consistent with PEP 8, but for our goal of
+complete unambiguity and uniformity it falls short on two counts: first, it
+does not mention the documentation; second, PEP 8 recommends "modules should
+explicitly declare the names in their public API using the ``__all__``
+attribute. Setting ``__all__`` to an empty list indicates that the module has no
+public API." SciPy allows modules to lack an ``__all__`` attribute, meaning a
+user and their tools must understand the nuances of the previous rules. Having
+an ``__all__`` attribute in every module is simpler, unambiguous, and better for
+introspection by both users and automated systems.
+
+Finally, let's consider what `Python typing guide
+<https://github.com/python/typing/blob/master/docs/source/libraries.rst#library-interface-public-and-private-symbols>`
+adds.:: 
+
+   - Symbols whose names begin with an underscore (but are not dunder names) are
+     considered private.
+   
+   - A module can expose an ``__all__`` symbol at the module level that provides
+     a list of names that are considered part of the interface. This overrides
+     all other rules above, allowing imported symbols or symbols whose names
+     begin with an underscore to be included in the interface.
+
+This is consistent with PEP 8 and SciPy, but it does not mention the
+documentation. The Python typing guide reinforces the idea that the ``__all__``
+attribute is the authoritative source of what is public vs internal.
+Importantly, the ``__all__`` attribute is a module-level attribute, so it only
+applies to the module in which it is defined. We will refer to this as "locally"
+public. This means that a module can define an ``__all__`` attribute but if the
+module itself is not public, then anything in the ``__all__`` attribute cannot
+be publicly accessed from outside the module.
```

Comment:
I'm not sure I understand this sentence. Does this imply that any private module should set `__all__ = []`, or does it mean that `__all__`, within a private module, defines an interface that's meant for internal use only ? (now that I wrote it down, the latter seems way more likely, but I think there's room to improve the phrasing)

Reply:
Yes, the latter. If you can't publicly navigate to a module, it doesn't matter what's "public" within that module. The full path must be public.
The issue with relying on this alone is what you discovered in that Issue (trying to find the one) where we have modules that look public (because they are not underscored) so it's possible to construct an import path that looks totally public, but actually isn't.

Reply:
>  If you can't publicly navigate to a module, it doesn't matter what's "public" within that module.

I think that's what confused me; since it doesn't matter, why should we care what in a private module's `__all__` ? Wouldn't it be simpler to recommend/rule that `__all__` must be set to `[]` in private modules ?

> The issue with relying on this alone is what you discovered in that Issue (trying to find the one)

For reference, I believe you meant to link https://github.com/astropy/astropy/issues/15666

Reply:
It might still be useful to be able to indicate in a private sub-module which things are meant to be importable in other parts of astropy, so ``__all__`` could be useful in that way rather than always setting to ``[]``.

Reply:
Agreed! This is demonstrated in lines 281-287

Reply:
In examples of good module layouts.

Reply:
@astrofrog, see https://github.com/astropy/astropy/pull/15109/files as an example doing what you're talking about. I agree, this is super useful.